File: java/matching-conformance/src/main/java/io/github/giovanicaprison/matching/conformance/Corpus.java
package io.github.giovanicaprison.matching.conformance;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.stream.Stream;

/**
 * The fixtures on disk.
 *
 * <p>They sit above both language trees, so neither owns the file that holds both to the same
 * behaviour. The directory is found by walking up rather than configured, so this works from a
 * module, from the repository root and from an IDE.
 */
public final class Corpus {

  private static final String DIRECTORY = "corpus";

  private Corpus() {}

  public static List<Fixture> fixtures() {
    return files().stream().map(FixtureParser::parse).toList();
  }

  public static List<Path> files() {
    try (Stream<Path> entries = Files.list(directory())) {
      return entries.filter(path -> path.toString().endsWith(".txt")).sorted().toList();
    } catch (final IOException e) {
      throw new UncheckedIOException("cannot list the corpus", e);
    }
  }

  public static Path directory() {
    Path candidate = Path.of("").toAbsolutePath();
    while (candidate != null) {
      final Path corpus = candidate.resolve(DIRECTORY);
      if (Files.isDirectory(corpus)) {
        return corpus;
      }
      candidate = candidate.getParent();
    }
    throw new IllegalStateException(
        "no " + DIRECTORY + " directory above " + Path.of("").toAbsolutePath());
  }
}

File: cpp/conformance/corpus.cpp
#include "conformance/corpus.hpp"

#include <algorithm>

#include "conformance/fixture_parser.hpp"

namespace io::github::giovanicaprison::matching::conformance {

std::filesystem::path corpusDirectory() {
  std::filesystem::path candidate = std::filesystem::current_path();
  while (true) {
    const std::filesystem::path corpus = candidate / "corpus";
    if (std::filesystem::is_directory(corpus)) {
      return corpus;
    }
    if (candidate == candidate.root_path()) {
      throw std::runtime_error("no corpus directory above " +
                               std::filesystem::current_path().string());
    }
    candidate = candidate.parent_path();
  }
}

std::vector<Fixture> corpusFixtures() {
  std::vector<std::filesystem::path> files;
  for (const auto& entry : std::filesystem::recursive_directory_iterator(corpusDirectory())) {
    if (entry.is_regular_file() && entry.path().extension() == ".txt") {
      files.push_back(entry.path());
    }
  }
  std::sort(files.begin(), files.end());
  std::vector<Fixture> fixtures;
  fixtures.reserve(files.size());
  for (const std::filesystem::path& file : files) {
    fixtures.push_back(parseFixture(file));
  }
  return fixtures;
}

}  // namespace io::github::giovanicaprison::matching::conformance
